### DISCUSSION TITLE
Automatically Export CSV/YAML - How?

### DISCUSSION THREAD

**Discussion by adparis99** [netbox-community/netbox#11098] [Q&A] [answered]
Hey there, 

One of our biggest concerns is that the NetBox server might be be unavailable at some point for any number of reasons, and as such we'd like to automatically export certain models such as IP addresses on a regular basis to a separate location. Is this doable via custom scripts? Can it be done through the API? I know the sample custom script exemplifies creating a CSV in a more manual way, but we'd like the export to perfectly mirror the All Data option if possible. Thanks in advance for your help!

**Comment by mtinberg** [comment, marked as answer]
Yeah there are a bunch of different ways you can do this


  *   Write a small REST client to the API to fetch all the data for the data models you want to back up and write JSON. Query with jq in an emergency
  *   Write an Ansible playbook using netbox.nb_lookup API client to query all the data and write out using | to_json or | to_yaml to files, query with jq
  *   Write a Netbox Script to do all the queries in Python using the Django ORM and return the result as the output, trigger the Script over REST and write the output to JSON, query with jq
  *   Export Templates can be attached to the data model, written in Jinja2, and triggered over the API ?export=TemplateName so you can get the data in whatever form you want, CSV, JSON, HTML, Markdown, man pages, etc.
  *   The Django manage.py may have some generic export function
  *   ?? other options




—

Mark Tinberg ***@***.***>

Division of Information Technology-Network Services

University of Wisconsin-Madison


________________________________
From: adparis99 ***@***.***>
Sent: Friday, December 2, 2022 4:11 PM
To: netbox-community/netbox ***@***.***>
Cc: Subscribed ***@***.***>
Subject: [netbox-community/netbox] Automatically Export CSV/YAML - How? (Discussion #11098)


Hey there,

One of our biggest concerns is that the NetBox server might be be unavailable at some point for any number of reasons, and as such we'd like to automatically export certain models such as IP addresses on a regular basis to separate location. Is this doable via custom scripts? Can it be done through the API? I know the sample custom script exemplifies creating a CSV in a more manual way, but we'd like the export to perfectly mirror the All Data option if possible. Thanks in advance for your help!

**Reply by adparis99** [reply]
Thanks Mark. For anyone else that might be looking for something similar in the future, I settled on the template below in order to generate an importable CSV backup:

`address,vrf,tenant,status,role,device,virtual_machine,interface,is_primary,dns_name,description,cf_dhcp_reservation,cf_mac_address
{% for ip in queryset -%}
{{ ip.address if ip.address is not none }},{{ ip.vrf if ip.vrf is not none }},{{ ip.tenant if ip.tenant is not none }},{{ ip.status if ip.status is not none }},{{ ip.role if ip.role is not none }},{{ ip.assigned_object.device.name if ip.assigned_object is not none }},{{ ip.virtual_machine if ip.virtual_machine is not none }},{{ ip.assigned_object.name if ip.assigned_object.name is not none }},{{ ip.is_primary if ip.is_primary is not none }},{{ ip.dns_name if ip.dns_name is not none }},{{ ip.description if ip.description is not none }},{{ ip.cf_dhcp_reservation if ip.cf_dhcp_reservation is not none }},{{ ip.cf_mac_address if ip.cf_mac_address is not none }}
{% endfor %}`

**Comment by MicrosoftJones** [comment]
https://github.com/netbox-community/netbox/discussions/11098#discussioncomment-4296807
